benchmark/core/Cabana_CommPerformance.cpp

```diff
@@ -348,20 +368,132 @@ void performanceTest( std::ostream& stream, const std::size_t num_particle,
                 comm_memory_space(), particles );
 
             s0 = Cabana::slice<0>( comm_particles );
-            Cabana::scatter( halo_fast, s0 );
+            Cabana::scatter( halo, s0 );
 
             s1 = Cabana::slice<1>( comm_particles );
-            Cabana::scatter( halo_fast, s1 );
+            Cabana::scatter( halo, s1 );
 
             s2 = Cabana::slice<2>( comm_particles );
-            Cabana::scatter( halo_fast, s2 );
+            Cabana::scatter( halo, s2 );
 
             s3 = Cabana::slice<3>( comm_particles );
-            Cabana::scatter( halo_fast, s3 );
+            Cabana::scatter( halo, s3 );
 
             Cabana::deep_copy( particles, comm_particles );
             halo_slice_scatter.stop( fraction );
         }
+
+        // Preallocated buffers.
+        // ---
+        double overallocation = 1.1;
+
+        // Run preallocated AoSoA buffer tests and time the ensemble.
+        auto comm_particles = Cabana::create_mirror_view_and_copy(
+            comm_memory_space(), particles );
+        auto send_buffer_aosoa = Cabana::gatherAllocateSendBuffer(
```

Comment:
I'm going to argue we don't need a separate interface for this and the `Halo` should just do this automatically. We should instead go for a `reserve()` and `shrinkToFit()` interface like we do elsewhere. The reason for this is that the `Halo` needs to be reconstructed anyway every time the particles move and get new ranks or new counts. We should therefore do the needed allocation there automatically. What I do want is instead of making a completely brand new halo object is to instead have a persistent halo with an `update()` function that takes in new particle counts/destination ranks.

Reply:
Switched back to draft after our discussion on how to combine theses persistent buffers into cleaner `Gather`/`Scatter` objects. I'll try to get that started soon